flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorTest.java

```diff
@@ -218,6 +230,67 @@ public void testNotifyCheckpointAbortedBeforeAsyncPhase() throws Exception {
 		assertEquals(0, subtaskCheckpointCoordinator.getAsyncCheckpointRunnableSize());
 	}
 
+	@Test
+	public void testDownstreamReceiveCancelCheckpointMarkerOnUpstreamAbortedInSyncPhase() throws Exception {
+		final OneInputStreamTaskTestHarness<String, String> testHarness =
+			new OneInputStreamTaskTestHarness<>(
+				OneInputStreamTask::new,
+				1, 1,
+				BasicTypeInfo.STRING_TYPE_INFO,
+				BasicTypeInfo.STRING_TYPE_INFO);
+
+		testHarness.setupOutputForSingletonOperatorChain();
+		StreamConfig streamConfig = testHarness.getStreamConfig();
+		streamConfig.setStreamOperator(new MapOperator());
+
+		testHarness.invoke();
+		testHarness.waitForTaskRunning();
+
+		TestTaskStateManager stateManager = new TestTaskStateManager();
+		MockEnvironment mockEnvironment = MockEnvironment.builder().setTaskStateManager(stateManager).build();
+		SubtaskCheckpointCoordinatorImpl subtaskCheckpointCoordinator = (SubtaskCheckpointCoordinatorImpl) new MockSubtaskCheckpointCoordinatorBuilder()
+			.setEnvironment(mockEnvironment)
+			.setUnalignedCheckpointEnabled(true)
+			.build();
+
+		final TestPooledBufferProvider bufferProvider = new TestPooledBufferProvider(Integer.MAX_VALUE, 4096);
```

Comment:
nit: only 1 buffer is enough for this case. `Integer.MAX_VALUE` might bring some memory concerns if we change the implementation of `TestPooledBufferProvider` future. E.g. if we allocate the buffer early during constructor based on the size.